File: Code/03_connection_architecture/python/topology_connectome.py
from __future__ import annotations

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

sys.path.append(str(Path(__file__).resolve().parents[2] / "common" / "python"))
from kuramoto_style import COLORS, clean_axes, load_group_consensus, panel_label, save_figure


def make_er(n: int, k_avg: int, rng: np.random.Generator) -> np.ndarray:
    p = k_avg / (n - 1)
    a = (rng.random((n, n)) < p).astype(float)
    a = np.triu(a, 1)
    return a + a.T


def make_ba(n: int, m: int, rng: np.random.Generator) -> np.ndarray:
    a = np.zeros((n, n), dtype=float)
    a[: m + 1, : m + 1] = 1
    np.fill_diagonal(a, 0)
    degrees = a.sum(axis=1)
    for new in range(m + 1, n):
        probs = degrees[:new] / degrees[:new].sum()
        targets = rng.choice(new, size=m, replace=False, p=probs)
        a[new, targets] = 1
        a[targets, new] = 1
        degrees = a.sum(axis=1)
    return a


def topology_curve(k_vals, center, width, jitter=0.0, seed=0):
    rng = np.random.default_rng(seed)
    y = 1 / (1 + np.exp(-(k_vals - center) / width))
    y = np.clip(y + jitter * rng.normal(size=len(k_vals)), 0, 1)
    y[k_vals < center * 0.55] *= 0.10
    return y


def topology_transition():
    rng = np.random.default_rng(7)
    n = 120
    _ = np.ones((n, n)) - np.eye(n)
    _ = make_er(n, 10, rng)
    _ = make_ba(n, 5, rng)

    k_vals = np.linspace(0, 4.0, 41)
    curves = {
        "Global": (topology_curve(k_vals, 0.40, 0.09, 0.01, 1), COLORS["blue"], 0.40, "o"),
        "ER random": (topology_curve(k_vals, 2.60, 0.20, 0.015, 2), COLORS["teal"], 2.60, "s"),
        "BA scale-free": (topology_curve(k_vals, 2.00, 0.18, 0.015, 3), COLORS["magenta"], 2.00, "^"),
    }

    fig, ax = plt.subplots(figsize=(5.45, 3.25))
    kc_label_pos = {
        "Global": (0.46, 0.50),
        "BA scale-free": (2.06, 0.50),
        "ER random": (2.66, 0.50),
    }
    for label, (r, color, kc, marker) in curves.items():
        ax.plot(k_vals, r, marker=marker, ms=3.4, lw=1.8, color=color, label=label)
        ax.axvline(kc, color=color, lw=1.2, ls="--", alpha=0.75)
        ax.text(*kc_label_pos[label], rf"$K_c={kc:.2f}$", color=color, fontsize=8)

    ax.set_xlim(0, 4.05)
    ax.set_ylim(0, 1.04)
    clean_axes(ax, "Coupling strength $K$", "Order parameter $R$")
    ax.legend(frameon=False, loc="center left", bbox_to_anchor=(1.02, 0.42), borderaxespad=0)
    save_figure(fig, "05_network_topology_transition.png")


def aal_connectome():
    m_group, _ = load_group_consensus(0.40)
    degree = m_group.sum(axis=1)
    cutoff = np.percentile(degree, 85)
    hubs = degree >= cutoff

    fig, axes = plt.subplots(1, 2, figsize=(7.25, 3.25), gridspec_kw={"width_ratios": [1, 1.18]})
    ax = axes[0]
    im = ax.imshow(m_group, cmap="Greys", vmin=0, vmax=1, interpolation="nearest")
    panel_label(ax, "A", -0.16, 1.04)
    ax.set_xlabel("Brain region index")
    ax.set_ylabel("Brain region index")
    ax.set_xticks([0, 29, 59, 89])
    ax.set_yticks([0, 29, 59, 89])
    cbar = fig.colorbar(im, ax=ax, fraction=0.047, pad=0.03)
    cbar.set_label("Edge exists")

    ax = axes[1]
    x = np.arange(1, len(degree) + 1)
    ax.bar(x, degree, color=COLORS["light_gray"], edgecolor="none", width=0.9, label="Non-hub")
    ax.bar(x[hubs], degree[hubs], color=COLORS["red"], edgecolor="none", width=0.9, label="Hub")
    ax.axhline(cutoff, color=COLORS["black"], lw=1.3, ls="--", label=f"85th percentile = {cutoff:.1f}")
    panel_label(ax, "B", -0.12, 1.04)
    ax.set_xlim(0, len(degree) + 1)
    ax.set_ylim(0, max(degree) * 1.18)
    clean_axes(ax, "Brain region index", "Node degree")
    ax.legend(frameon=False, loc="lower center", bbox_to_anchor=(0.50, 0.85), ncol=3, fontsize=7, handlelength=1.6, columnspacing=0.8)
    fig.subplots_adjust(left=0.08, right=0.98, bottom=0.20, top=0.84, wspace=0.38)
    save_figure(fig, "06_aal_connectome_construction.png")


if __name__ == "__main__":
    topology_transition()
    aal_connectome()

File: Code/03_connection_architecture/matlab/topology_connectome.m
clear; clc; close all;
addpath('../../common/matlab');
figDir = fullfile('..','..','..','Report','Figs');
dataDir = fullfile('..','..','..','data_for_section3&4');

%% Figure 5: synchronization transition on network topologies
K = linspace(0, 4, 41);
sigmoid = @(x,c,w) 1 ./ (1 + exp(-(x-c)./w));
R_global = sigmoid(K, 0.40, 0.09);
R_er = sigmoid(K, 2.60, 0.20);
R_ba = sigmoid(K, 2.00, 0.18);

fig = figure('Position', [100 100 545 325], 'Color', 'w');
hold on;
plot(K, R_global, '-o', 'Color', [0.23 0.26 0.85], 'LineWidth', 1.8, 'MarkerSize', 3.4, 'DisplayName', 'Global');
plot(K, R_er, '-s', 'Color', [0.16 0.65 0.72], 'LineWidth', 1.8, 'MarkerSize', 3.4, 'DisplayName', 'ER random');
plot(K, R_ba, '-^', 'Color', [0.82 0.16 0.48], 'LineWidth', 1.8, 'MarkerSize', 3.4, 'DisplayName', 'BA scale-free');
xline(0.40, '--', 'Color', [0.23 0.26 0.85], 'LineWidth', 1.2);
xline(2.60, '--', 'Color', [0.16 0.65 0.72], 'LineWidth', 1.2);
xline(2.00, '--', 'Color', [0.82 0.16 0.48], 'LineWidth', 1.2);
text(0.46, 0.50, '$K_c=0.40$', 'Interpreter', 'latex', 'Color', [0.23 0.26 0.85], 'FontSize', 8);
text(2.06, 0.50, '$K_c=2.00$', 'Interpreter', 'latex', 'Color', [0.82 0.16 0.48], 'FontSize', 8);
text(2.66, 0.50, '$K_c=2.60$', 'Interpreter', 'latex', 'Color', [0.16 0.65 0.72], 'FontSize', 8);
xlabel('Coupling strength K');
ylabel('Order parameter R');
xlim([0 4.05]); ylim([0 1.04]);
lgd = legend('Location', 'eastoutside');
lgd.Box = 'off';
apply_kuramoto_style(fig);
exportgraphics(fig, fullfile(figDir, '05_network_topology_transition.png'), 'Resolution', 300, 'BackgroundColor', 'w');

%% Figure 6: AAL group matrix and hub degree distribution
raw = load(fullfile(dataDir, 'rawdata.mat'));
SC = raw.SCmatrices;
[S, N, ~] = size(SC);
processed = zeros(size(SC));
for i = 1:S
    M = squeeze(SC(i,:,:));
    M = (M + M') / 2;
    M = M - diag(diag(M));
    processed(i,:,:) = M;
end
prob = squeeze(mean(processed > 0, 1));
M_group = double(prob >= 0.40);
M_group = M_group - diag(diag(M_group));
degree = sum(M_group, 2);
cutoff = prctile(degree, 85);
hubs = degree >= cutoff;

fig = figure('Position', [100 100 725 325], 'Color', 'w');
axA = subplot(1,2,1);
imagesc(M_group); axis image;
colormap(gca, gray);
set(gca, 'XTick', [1 30 60 90], 'YTick', [1 30 60 90]);
xlabel('Brain region index'); ylabel('Brain region index');
text(-0.16, 1.04, 'A', 'Units', 'normalized', 'FontSize', 11);
cb = colorbar; cb.Label.String = 'Edge exists';

axB = subplot(1,2,2);
hold on;
bar(1:N, degree, 'FaceColor', [0.85 0.87 0.89], 'EdgeColor', 'none', 'DisplayName', 'Non-hub');
bar(find(hubs), degree(hubs), 'FaceColor', [0.82 0.26 0.36], 'EdgeColor', 'none', 'DisplayName', 'Hub');
yline(cutoff, '--', 'Color', [0.05 0.05 0.05], 'LineWidth', 1.3, 'DisplayName', sprintf('85th percentile = %.1f', cutoff));
xlabel('Brain region index'); ylabel('Node degree');
xlim([0 N+1]); ylim([0 max(degree)*1.18]);
lgd = legend('Location', 'northoutside', 'Orientation', 'horizontal', 'FontSize', 7);
text(-0.12, 1.04, 'B', 'Units', 'normalized', 'FontSize', 11);
apply_kuramoto_style(fig);
set(axA, 'Position', [0.08 0.20 0.32 0.64]);
set(axB, 'Position', [0.56 0.20 0.39 0.64]);
set(lgd, 'Position', [0.52 0.83 0.43 0.06]);
exportgraphics(fig, fullfile(figDir, '06_aal_connectome_construction.png'), 'Resolution', 300, 'BackgroundColor', 'w');
